pkg/helpers/reconcile.go

```diff
@@ -0,0 +1,350 @@
+		// Delete the HelmRelease(s) - parent and subchart(s)
+		// Lookup charts using the label selector.
+		// Example: chart=kafka-dev,heritage=orkestra,owner=dev, where chart=<top-level-chart>
+		helper.Info("Remediating the applicationgroup with helmrelease failure status")
+		for _, app := range helper.Instance.Status.Applications {
+			switch meta.GetResourceCondition(&app.ChartStatus, meta.ReadyCondition).Reason {
+			case fluxhelmv2beta1.InstallFailedReason, fluxhelmv2beta1.UpgradeFailedReason, fluxhelmv2beta1.UninstallFailedReason,
+				fluxhelmv2beta1.ArtifactFailedReason, fluxhelmv2beta1.InitFailedReason, fluxhelmv2beta1.GetLastReleaseFailedReason:
+				listOption := client.MatchingLabels{
+					workflow.OwnershipLabel: helper.Instance.Name,
+					workflow.HeritageLabel:  workflow.Project,
+					workflow.ChartLabelKey:  app.Name,
+				}
+				helmReleases := fluxhelmv2beta1.HelmReleaseList{}
+				err = helper.List(ctx, &helmReleases, listOption)
+				if err != nil {
+					helper.Error(err, "failed to find generated HelmRelease instances")
+					return reconcile.Result{}, nil
+				}
+
+				err = helper.rollbackFailedHelmReleases(ctx, helmReleases.Items)
+				if err != nil {
+					helper.Error(err, "failed to rollback failed HelmRelease instances")
+					return reconcile.Result{}, nil
+				}
+			}
+		}
+	}
+	// mark the object as requiring rollback so that we can rollback
+	// to the previous versions of all the applications in the ApplicationGroup
+	// using the last successful spec
+	helper.Info("Rolling back to last successful application group spec")
+	helper.Instance.Spec = *helper.Instance.GetLastSuccessful()
+	if err := helper.Patch(ctx, helper.Instance, patch); err != nil {
+		helper.Error(err, "failed to update ApplicationGroup instance while rolling back")
+		return ctrl.Result{}, err
+	}
+	return reconcile.Result{RequeueAfter: v1alpha1.DefaultProgressingRequeue}, nil
+}
+
+func (helper *ReconcileHelper) Reverse(ctx context.Context) (ctrl.Result, error) {
+	if workflow := helper.GetWorkflow(ctx); workflow != nil {
+		helper.Info("cleaning up the workflow object")
+		if err := helper.cleanupWorkflow(ctx, helper.Logger, workflow); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+	return reconcile.Result{}, nil
+}
+
+func (helper *ReconcileHelper) GetWorkflow(ctx context.Context) *v1alpha12.Workflow {
+	wfs := v1alpha12.WorkflowList{}
+	listOption := client.MatchingLabels{
+		workflow.OwnershipLabel: helper.Instance.Name,
+		workflow.HeritageLabel:  workflow.Project,
+	}
+	_ = helper.List(ctx, &wfs, listOption)
+
+	if len(wfs.Items) != 0 {
+		return &wfs.Items[0]
+	}
+	return nil
+}
+
+func (helper *ReconcileHelper) reconcileApplications() error {
+	stagingDir := helper.RegistryOptions.TargetDir + "/" + helper.RegistryOptions.StagingRepoName
+
+	// Init the application status every time we re-reconcile the applications
+	InitAppStatus(helper.Instance)
```

Comment:
lowercase this function as it is not used elsewhere outside of the helpers package